src/utils/list_cpu_features.c

```diff
@@ -30,7 +36,176 @@
 #include "cpuinfo_ppc.h"
 #endif
 
-static void PrintEscapedAscii(const char* str) {
+// Design principles
+// -----------------
+// We build a tree structure containing all the data to be displayed.
+// Then depending on the output type (text or json) we walk the tree and display
+// the data accordingly.
+
+// We use a bump allocator to allocate strings and nodes of the tree,
+// Memory is not intented to be reclaimed.
+typedef struct {
+  char* ptr;
+  size_t size;
+} BumpAllocator;
+
+// Allocate a buffer of size `size`.
+static BumpAllocator BA_Create(size_t size) {
+  char* const ptr = (char*)malloc(size);
+  BumpAllocator BA;
+  if (ptr) BA = (BumpAllocator){.ptr = ptr, .size = size};
+  return BA;
+}
+
+// Update the available memory left in the BumpAllocator.
+static void* BA_Bump(BumpAllocator* BA, size_t size) {
+  assert(BA->size >= size);
+  void* ptr = BA->ptr;
+  BA->size -= size;
+  BA->ptr += size;
+  return ptr;
+}
+
+// The type of the nodes in the tree.
+typedef enum {
+  TNT_INVALID,
+  TNT_INT,
+  TNT_MAP,
+  TNT_MAP_ENTRY,
+  TNT_ARRAY,
+  TNT_ARRAY_ELEMENT,
+  TNT_STRING,
+} TreeValueType;
+
+// The node in the tree.
+typedef struct TreeValue {
+  TreeValueType type;
+  unsigned integer;
+  const char* string;
+  struct TreeValue* value;
+  struct TreeValue* next;
+} TreeValue;
```

Comment:
nit: It may make more sense to call this a Dictionary than a Tree, since it doesn't have the regular left/right children nodes and performance characteristics of trees.